### DISCUSSION TITLE
Are errors from a `Guard` outcome accessible through a `Catcher` fn?

### DISCUSSION THREAD

**Discussion by akappel** [rwf2/Rocket#1886] [Questions] [answered]
Hey everyone 👋 I posed this question in the IRC but was unable to stick around to see if it was answered:

I may be misunderstanding, but if a guard returns a `Failure` outcome with a custom error type, are we not able to access that error in a `Catcher` fn? Is our only option to use a `Result<S, E>` in the request handler?

Here's an example:
```rust
#[macro_use] extern crate rocket;

use rocket::{fs::NamedFile, http::Status, request::FromRequest, request::Outcome, response::status::NotFound};

#[get("/")]
async fn index(_guard: MyGuard) -> Result<NamedFile, NotFound<String>> {
    NamedFile::open("templates/index.html")
        .await
        .map_err(|e| NotFound(e.to_string()))
}

#[launch]
fn rocket() -> _ {
    rocket::build()
        .mount("/", routes![index])
        .register("/", catchers![catch_401])
}

#[catch(401)]
fn catch_401() -> &'static str {
    "ACCESS DENIED"
}

#[derive(Debug)]
enum MyError {
    ChromeError(&'static str),
}

struct MyGuard;

#[rocket::async_trait]
impl<'r> FromRequest<'r> for MyGuard {
    type Error = MyError;

    async fn from_request(request: &'r rocket::Request<'_>) -> Outcome<Self, Self::Error> {
        let user_agent = String::from(request.headers().get_one("User-Agent").unwrap());
        if user_agent.contains("Chrome") {
            return Outcome::Failure((Status::Unauthorized, MyError::ChromeError("ACCESS DENIED")));
        }

        Outcome::Success(Self)
    }
}
```

The docs make no mention of whether
```rust
#[catch(401)]
fn catch_401() -> &'static str {
    "ACCESS DENIED"

    // can I log MyError here?
}
```
can use `MyError`. The only reference to using the error is in the handler's [outcome reference](https://api.rocket.rs/master/rocket/outcome/index.html), by using a result as a param:
> In Rocket, a Failure generally means that a request is taken out of normal processing. The request is then given to the catcher corresponding to some status code. Users can catch failures by requesting a type of Result<S, E> or Option in request handlers.

Maybe it's not a good idea that the catcher should have access to that error content? Or is it that the `catch` attribute doesn't consider an error being provided? Thanks!

**Comment by jebrosen** [comment, marked as answer]
This is a desired feature (see #749), which unfortunately required or seemed to require new features in the rust language.

**Reply by akappel** [reply]
Thanks for the background @jebrosen. I'll work around it!
